src/ocean/meta/values/VisitValue.d

```diff
@@ -0,0 +1,183 @@
+
+module ocean.meta.values.VisitValue;
+
+import ocean.meta.traits.Basic;
+import ocean.meta.types.Arrays;
+import ocean.meta.types.Typedef;
+
+/*******************************************************************************
+
+    Recursively iterates over `value` fields and calls `visitor.visit` for each
+    one of those. `visitor.visit` is expected to be a template with the
+    following signature:
+
+    `bool visit ( T ) ( T* value )`
+
+    The return value indicates whether the value should be recursed into or
+    not.
+
+    Example:
+
+    ---
+    struct SumIntegersVisitor
+    {
+        /// Visitor can have any internal state it may need
+        long sum;
+
+        /// Returns: 'true' if recursion shall continue for this type
+        bool visit ( T ) ( T* value )
+        {
+            static if (isIntegerType!(T))
+                sum += *value;
+
+            return true;
+        }
+    }
+
+    SumIntegersVisitor visitor;
+    visitValue(some_value, visitor);
+    ---
+
+    Params:
+        value = root value to visit
+        visitor = struct providing templated `visit` method accepting single
+            argument which will be a pointer to visited element
+
+*******************************************************************************/
+
+public void visitValue ( T, Visitor ) ( ref T value, ref Visitor visitor )
+{
+    VisitImpl!(Visitor) impl;
+    impl.visitor = &visitor;
+    impl.visitAll(&value);
+}
+
+/*******************************************************************************
+
+    Implementation of visiting logic. Some types are currently not supported
+    because it is not clear what would recursing mean for them.
+
+    Params:
+        Visitor = user-provided struct implementing templated `visit` method
+            to be called for individual values
+
+*******************************************************************************/
+
+private struct VisitImpl ( Visitor )
+{
+    /// Refers to the persistent instance of visitor struct in case
+    /// implementation / may need internal state
+    Visitor* visitor;
+
+    /***************************************************************************
+
+        Main recursive type/value visiting implementation
+
+        Params:
+            value = pointer to currently processed value. Pointer is used
+                instead of a reference because D1 does not allow defining
+                a reference to a static array
+
+    ***************************************************************************/
+
+    private void visitAll ( T ) ( T* value )
+    {
+        auto deeper = this.visitor.visit(value);
+
+        if (!deeper)
+            return;
+
+        static if (isPrimitiveType!(T))
+        {
+            // do nothing, already processed
+        }
+        else static if (isTypedef!(T))
+        {
+            auto reinterp = cast(TypedefBaseType!(T)*) value;
+            this.visitAll(reinterp);
```

Comment:
Why not `this.visitAll(cast(TypedefBaseType!(T)*) value);` like for classes below? (Same for enums below.)

Reply:
Because I wanted to stress the fact this is reinterpret cast and not a "smart" D cast.

Reply:
(I switched class version to your proposal so it is more consistent now)